src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs

```diff
@@ -284,15 +301,31 @@ private bool CData()
             }
             else
             {
-                var tagName = String.Empty;
-                if (At(HtmlSymbolType.Text))
+                var tagName = string.Empty;
+                HtmlSymbol bangSymbol = null;
+
+                if (At(HtmlSymbolType.Bang))
+                {
+                    bangSymbol = CurrentSymbol;
+
+                    var nextSymbol = Lookahead();
+
+                    if (nextSymbol.Type == HtmlSymbolType.Text)
+                    {
+                        tagName = nextSymbol.Content;
+                    }
+                }
+                else if (At(HtmlSymbolType.Text))
                 {
                     tagName = CurrentSymbol.Content;
                 }
+
                 var matched = RemoveTag(tags, tagName, tagStart);
 
                 if (tags.Count == 0 &&
-                    String.Equals(tagName, SyntaxConstants.TextTagName, StringComparison.OrdinalIgnoreCase) &&
+                    // text tags cannot be targeted by TagHelpers and therefore can't be escaped with '!'.
```

Comment:
unnecessary error.  treat `<!text>` the same as `<text>`.  i.e. use `!` to mean "never a tag helper" even if the result is something other than a tag.